chainstate/tx-verifier/src/transaction_verifier/mod.rs

```diff
@@ -589,48 +677,197 @@ where
                 }
             };
         }
 
+        self.disconnect_pos_accounting_outputs(tx_source, tx)?;
+
+        self.disconnect_tokens_accounting_outputs(tx_source, tx)?;
+
+        Ok(())
+    }
+
+    fn disconnect_pos_accounting_outputs(
+        &mut self,
+        tx_source: TransactionSource,
+        tx: &Transaction,
+    ) -> Result<(), ConnectTransactionError> {
         // apply undos to accounting
         let block_undo_fetcher = |id: Id<Block>| {
             self.storage
                 .get_accounting_undo(id)
                 .map_err(|_| ConnectTransactionError::TxVerifierStorage)
         };
-        let undos = self.accounting_block_undo.take_tx_undo(
+        let undos = self.pos_accounting_block_undo.take_tx_undo(
             &tx_source,
             &tx.get_id(),
             block_undo_fetcher,
         )?;
         if let Some(undos) = undos {
             undos.into_inner().into_iter().try_for_each(|undo| {
-                self.accounting_delta_adapter.operations(tx_source).undo(undo)
+                self.pos_accounting_adapter.operations(tx_source).undo(undo)
             })?;
         }
 
         Ok(())
     }
 
+    fn connect_tokens_outputs(
+        &mut self,
+        tx_source: &TransactionSourceForConnect,
+        tx: &Transaction,
+    ) -> Result<(), ConnectTransactionError> {
+        // Check if v0 tokens are allowed to be used at this height
+        let latest_token_version = self
+            .chain_config
+            .as_ref()
+            .chainstate_upgrades()
+            .version_at_height(tx_source.expected_block_height())
+            .1
+            .token_issuance_version();
+
+        match latest_token_version {
+            TokenIssuanceVersion::V0 => { /* ok */ }
+            TokenIssuanceVersion::V1 => {
+                let has_v0_tokens = tx.outputs().iter().any(|output| match output {
+                    TxOutput::Transfer(output_value, _)
+                    | TxOutput::Burn(output_value)
+                    | TxOutput::LockThenTransfer(output_value, _, _) => match output_value {
+                        OutputValue::Coin(_) | OutputValue::TokenV1(_, _) => false,
+                        OutputValue::TokenV0(_) => true,
+                    },
+                    TxOutput::CreateStakePool(_, _)
+                    | TxOutput::ProduceBlockFromStake(_, _)
+                    | TxOutput::CreateDelegationId(_, _)
+                    | TxOutput::DelegateStaking(_, _)
+                    | TxOutput::TokensOp(_) => false,
+                });
+                ensure!(
+                    !has_v0_tokens,
+                    ConnectTransactionError::TokensError(
+                        TokensError::DeprecatedTokenIssuanceVersion(
+                            tx.get_id(),
+                            TokenIssuanceVersion::V0,
+                        ),
+                    )
+                );
+            }
+            _ => unreachable!(),
```

Comment:
Can we make latest_token_version somehow an enum to get rid of this unreachable?